update (passive_anticheat): Monthly alignment

acidmanifesto opened the pull request:
**Changes proposed**:

- New Conf: Anticheat.SpeedLimitTolerance
Create a conf to establish a speed limit tolerance over server rate set speed.
- New Conf: Anticheat.BG.StartAreaTeleport
BG Counter measure for starter zone BG's.
This is done so we can ignore minor violations that are not false positives such as going 1 or 2 over the speed limit
- Fine Tune BG AV from false detection on horde side.
- Add additional anticheat helper

**Target branch(es)**: 335-passive_anticheat

**Issues addressed**: Fixes #

**Tests performed**: (Does it build, tested in-game, etc)

**Known issues and TODO list**:

- [ ] 
- [ ]

acidmanifesto committed:
update (passive_anticheat): Monthly alignment

acidmanifesto committed:
update (Detection):BG Fine Tune

acidmanifesto committed:
Update (detection): False Flag Advance jump

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -237,22 +237,32 @@ void AnticheatMgr::SpeedHackDetection(Player* player, MovementInfo movementInfo)
 
-    // We did the (uint32) cast to accept a margin of tolerance
+    // we create a diff speed in uint32 for further precision checking to avoid legit fall and slide
+    uint32 diffspeed = clientSpeedRate - speedRate;
```
Is this handling cases where clientSpeedRate is less than speedRate, so diffspeed is a negative number stored in a uint32 ?
I think you need to write the `if (diffspeed >= _assignedspeeddiff)` check with using + instead of -, so `if (clientSpeedRate >= _assignedspeeddiff + speedRate)`

acidmanifesto replied:
sorry. been a long few days, im just trying to make sure i understand the math in this.
so changeuint32 diffspeed = clientSpeedRate - speedRate; to
```
    uint32 diffspeed = clientSpeedRate + speedRate;
```
and adjust the if ((diffspeed >= _assignedspeeddiff) to
```
if ((diffspeed >= _assignedspeeddiff + speedRate)
```
Correct?

jackpoz replied:
> check with using + instead of -

I meant you should never use the - operator with 2 uint32 numbers where you are not sure if the result will be a positive number. uint32 is "unsigned int", "unsigned" meaning it doesn't handle negative numbers.

so, as you cannot use the - sign, you shouldn't try to get "diffspeed" at all, just do the math using + signs only.
so instead of 
```
if (diffspeed >= _assignedspeeddiff)
```
use
```
if (clientSpeedRate >= _assignedspeeddiff + speedRate)
```
which is the same logically from a math point of view, but operates in the range of uint32 (which is [0, 4 billions]

Let me show the math :P
```
uint32 diffspeed = clientSpeedRate - speedRate;
if (diffspeed >= _assignedspeeddiff)
```
```
if (clientSpeedRate - speedRate >= _assignedspeeddiff)
```
```
if (clientSpeedRate >= _assignedspeeddiff + speedRate)
```

acidmanifesto replied:
Math is hard :(
Haha
I will make adjustments this week. Thanks.

jackpoz replied:
btw if you don't think the speed will ever go above 2 billions ( which it should not), then you can just use `int32` instead of `uint32`

acidmanifesto replied:
> btw if you don't think the speed will ever go above 2 billions ( which it should not), then you can just use `int32` instead of `uint32`

at the moment we got a mix match with int and uint if i changed it to int32 for the diffspeed, issue being obvious with the clientSpeedRate needing to be uint32.

acidmanifesto committed:
update (core\anticheat): Additional Helper

add anticheat helper here to avoid false hits if relog during flight path travel

acidmanifesto committed:
Feat: (countermeasure) BG Countermeasure

If a player cheats by teleporting outside of the start spot before the BG starts, a counter measure will applied and log (if WriteLog is enabled) and send the player to a void spot for a few seconds and return them back in the team start spot

acidmanifesto committed:
add (logs): additional Counter Measure Logs added

acidmanifesto committed:
Adjustments

Update AnticheatMgr.cpp

Adjustments

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -228,31 +232,51 @@ void AnticheatMgr::SpeedHackDetection(Player* player, MovementInfo movementInfo)
+    // We did the (uint32) cast to accept a margin of tolerance for seasonal spells and buffs such as sugar rush
     // We check the last MovementInfo for the falling flag since falling down a hill and sliding a bit triggered a false positive
-    if ((clientSpeedRate > speedRate * 1.05f) && !m_Players[key].GetLastMovementInfo().HasMovementFlag(MOVEMENTFLAG_FALLING))
+    if ((diffspeed >= _assignedspeeddiff + speedRate) && !m_Players[key].GetLastMovementInfo().HasMovementFlag(MOVEMENTFLAG_FALLING))
```
```suggestion
    if ((clientSpeedRate >= _assignedspeeddiff + speedRate) && !m_Players[key].GetLastMovementInfo().HasMovementFlag(MOVEMENTFLAG_FALLING))
```

jackpoz replied:
```suggestion
    if ((clientSpeedRate >= _assignedspeeddiff + speedRate) && !m_Players[key].GetLastMovementInfo().HasMovementFlag(MOVEMENTFLAG_FALLING))
```

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -228,31 +232,51 @@ void AnticheatMgr::SpeedHackDetection(Player* player, MovementInfo movementInfo)
 
-    // We did the (uint32) cast to accept a margin of tolerance
+    // we create a diff speed in uint32 for further precision checking to avoid legit fall and slide
+    uint32 diffspeed = clientSpeedRate + speedRate;
```
```suggestion
```

acidmanifesto committed:
Update src/server/game/Anticheat/AnticheatMgr.cpp

Co-authored-by: Giacomo Pozzoni <giacomopoz@gmail.com>

acidmanifesto committed:
Update src/server/game/Anticheat/AnticheatMgr.cpp

Co-authored-by: Giacomo Pozzoni <giacomopoz@gmail.com>

acidmanifesto committed:
Update AnticheatMgr.cpp

acidmanifesto committed:
update

jackpoz commented:
is this ready to be merged ?

acidmanifesto commented:
> is this ready to be merged ?

Not yet. Havnt had time to update the logging

acidmanifesto committed:
Update AnticheatMgr.cpp

acidmanifesto committed:
Update AnticheatMgr.cpp

acidmanifesto committed:
Update AnticheatMgr.cpp

acidmanifesto committed:
Update AnticheatMgr.cpp

acidmanifesto committed:
null check

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.cpp:
```diff
@@ -228,31 +232,50 @@ void AnticheatMgr::SpeedHackDetection(Player* player, MovementInfo movementInfo)
+    if ((clientSpeedRate >= _assignedspeeddiff + speedRate) && !m_Players[key].GetLastMovementInfo().HasMovementFlag(MOVEMENTFLAG_FALLING))
     {
-        if (!player->CanTeleport())
+        if ((clientSpeedRate >= _assignedspeeddiff + speedRate) && !m_Players[key].GetLastMovementInfo().HasMovementFlag(MOVEMENTFLAG_FALLING))
```
isn't this check the same as the one 2 lines above ?

jackpoz commented on src/server/game/Anticheat/AnticheatMgr.h:
```suggestion
        uint32 m_MapId = uint32(-1);
```

acidmanifesto committed:
added code review suggestions

added code review suggestion per @jackpoz
Thank you.

Co-Authored-By: Giacomo Pozzoni <giacomopoz@gmail.com>
